File: problems/2025/09-18/next-greater-element-circular/solution.py
from typing import List

def next_greater(arr: List[int]) -> List[int]:
    n = len(arr)
    ans = [-1] * n
    stack: List[int] = []  # indices with decreasing values
    for i in range(2 * n - 1, -1, -1):
        j = i % n
        while stack and arr[stack[-1]] <= arr[j]:
            stack.pop()
        if i < n and stack:
            ans[j] = arr[stack[-1]]
        stack.append(j)
    return ans

if __name__ == "__main__":
    print(next_greater([1,3,2,4]))        # [3,4,4,-1]
    print(next_greater([0,2,3,1,1]))      # [2,3,-1,2,2]

File: problems/2025/09-18/next-greater-element-circular/solution.js
function nextGreater(arr) {
  const n = arr.length;
  const ans = Array(n).fill(-1);
  const st = []; // stack of indices
  for (let i = 2 * n - 1; i >= 0; i--) {
    const j = i % n;
    while (st.length && arr[st[st.length - 1]] <= arr[j]) st.pop();
    if (i < n && st.length) ans[j] = arr[st[st.length - 1]];
    st.push(j);
  }
  return ans;
}

console.log(nextGreater([1,3,2,4]));      // [3,4,4,-1]
console.log(nextGreater([0,2,3,1,1]));    // [2,3,-1,2,2]

